Source/Core/Duality/Resources/RectAtlas.cs

```diff
@@ -0,0 +1,497 @@
+			if (index >= this.items.Count)
+				throw new ArgumentOutOfRangeException("index", "The insertion index was greater than the size number of rects in the atlas");
+
+			if (this.tags != null)
+			{
+				List<string> emptyKeys = new List<string>();
+
+				// Adjust tagged indices to account for the removal of this item
+				foreach (var kvp in this.tags)
+				{
+					List<int> indexList = kvp.Value;
+					for (int i = 0; i < indexList.Count; i++)
+					{
+						// Index of rect being removed - remove from tag list
+						if (indexList[i] == index)
+						{
+							indexList.RemoveAt(i);
+							i--;
+						}
+						// Index after rect being remove - update it to new index
+						else if (indexList[i] > index)
+						{
+							indexList[i]--;
+						}
+					}
+
+					if (indexList.Count == 0)
+						emptyKeys.Add(kvp.Key);
+				}
+
+				// Remove any list of tagged indices that are now empty
+				foreach (string emptyKey in emptyKeys)
+					this.tags.Remove(emptyKey);
+			}
+
+			this.items.RemoveAt(index);
+		}
+
+		/// <summary>
+		/// Returns all rects with the given tag or an empty enumerable if no rects
+		/// with the given tag exist. The returned rects will be in index order.
+		/// </summary>
+		/// <param name="tag">The tag of the rects to look for</param>
+		/// <returns>
+		/// The rects with the given tag or an empty enumerable if no rects with the given tag exist.
+		/// </returns>
+		public IEnumerable<Rect> GetTaggedRects(string tag)
+		{
+			IEnumerable<int> indices = this.GetTaggedIndices(tag);
+			if (!indices.Any()) return Enumerable.Empty<Rect>();
+			return indices.Select(i => this[i]);
+		}
+
+		/// <summary>
+		/// Returns all the indices of all rects with the given tag or an empty enumerable if no rects
+		/// with the given tag exist.
+		/// </summary>
+		/// <param name="tag">The tag of the rects to look for.</param>
+		/// <returns>
+		/// The indices of rects with the given tag or an empty enumerable
+		/// if no rects with the given tag exist.
+		/// </returns>
+		public IEnumerable<int> GetTaggedIndices(string tag)
+		{
+			List<int> indices;
+			if (string.IsNullOrEmpty(tag))
+			{
+				indices = new List<int>();
+				for (int i = 0; i < this.items.Count; i++)
+				{
+					if (string.IsNullOrEmpty(this.items[i].Tag))
+						indices.Add(i);
+				}
+			}
+			else if (this.tags == null || !this.tags.TryGetValue(tag, out indices))
+			{
+				return Enumerable.Empty<int>();
+			}
+
+			return indices;
+		}
+
+		/// <summary>
+		/// Sets the tag associated with the given index.
+		/// </summary>
+		/// <param name="index">The index within the atlas to find the tag for.</param>
+		/// <param name="tag">The tag to associate with the given index.</param>
+		/// <returns>The tag of the given rect atlas index</returns>
+		/// <exception cref="ArgumentOutOfRangeException">
+		/// The given index was outside the range of the number of elements in the atlas.
+		/// </exception>
+		public void SetTag(int index, string tag)
```

Comment:
Why is settings tags and pivots separate from changing rect areas? 

Assuming we have item-based indexers instead of rect-based ones (see comments above), we could scratch all specialized API and use the indexer we have instead.

Current:
```c#
public Rect this[int index] { get; set; }

public string GetTag(int index);
public void SetTag(int index, string tag);
public void SetTags(string tag, int[] indices);

public Vector2 GetPivot(int index);
public void SetPivot(int index, Vector2 pivot);
```

Item-based:
```c#
public RectAtlasItem this[int index] { get; set; }
```

Depending on the most common use cases, we can always add convenience API later on - or leave it to the user via extension methods, since the basic functionality is there.